proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,641 @@
+* Ability to design better APIs (good error messages, no proxy types, no ambiguous types)
+* Prepare the compiler internals for further work on dependent types
+
+Proposed Change Specification
+-----------------------------
+
+1. Add a new language extension, ``VisibleForAll``.
+
+2. When ``VisibleForAll`` is in effect, lift the restriction that the ``forall a
+   ->`` quantifier cannot be used in types of terms.
+
+3. In types of terms, ``forall a ->`` is an irrelevant quantifier.
+
+4. Extend the term-level syntax with several constructs that previously could
+   only occur at the type level:
+
+   * Function arrows: ``a -> b``
+   * Multiplicity-polymorphic function arrows: ``a %m -> b`` (under ``-XLinearTypes``)
+   * Constraint arrows: ``a => b``
+   * Universal quantification: ``forall a. b``
+   * Visible universal quantification: ``forall a -> b``.
+
+   We will call them "types-in-terms". They are subject to the following rules:
+
+   * Grammatically, their constituents are terms,
+     not types::
+
+                proposed grammar:                 as opposed to:
+         +-------------------------------+----------------------------------+
+         |                               |                                  |
+         |  exp -> ...                   |    exp -> ...                    |
+         |       | exp -> exp            |         | type -> type           |
+         |       | exp => exp            |         | type => type           |
+         |       | forall bndrs. exp     |         | forall bndrs. type     |
+         |       | forall bndrs -> exp   |         | forall bndrs -> type   |
+         |                               |                                  |
+         +-------------------------------+----------------------------------+
+
+     This is a necessity to avoid parsing conflicts. As a consequence, the
+     ``'`` symbol signifies Template Haskell name quotation rather than
+     ``DataKinds`` promotion, and ``*`` is treated as an infix operator
+     regardless of ``-XStarIsType``.
+
+   * In name resolution, we switch to a type-level context when renaming
+     types-in-terms, with type-level identifiers having priority.
+
+   * In type-checking, types-in-terms are either rejected or mapped to types
+     (see "T2T-Mapping" below).
+
+5. Extend the name resolution rules: when looking up a term-level identifier
+   fails, look for a type-level identifier as fallback. As the result, a
+   term-level expression can also contain type variables and type constructors,
+   which will be either rejected or mapped to types (see "T2T-Mapping" below).
+
+6. In a term-level context, a reference to an ambiguous name (such as
+   ``Proxy``) is always resolved in favor of the data constructor, never the
+   type constructor.
+
+7. In a function application ``f e`` for some expressions ``f`` and ``e``, the
+   type of ``f`` affects neither parsing nor of ``e``.
```

Comment:
I think there is a word missing here, probably "renaming"

Reply:
Thanks!